### PR TITLE
Expose MMR root through runtime API - use it in BEEFY client

### PR DESCRIPTION
**Pull request by acatangiu**
Adds new `mmr_root()` API to `pallet-mmr` ; **doesn't** change runtime storage.
BEEFY client then uses this API as a backup to `MmrRoot` header digest logs.
Enables removal of `MmrRoot` header digest entries in the future => lighter headers.

- `beefy-gadget`: allow custom runtime api provider
- `beefy-gadget`: use mock runtime api in tests
- `pallet-mmr`: expose MMR root from state through runtime API
- `beefy-gadget`: get MMR root from runtime state
- `pallet-beefy-mmr`: remove MMR root from header digests

##

Moves MMR primitives from `frame/merkle-mountain-range/primitives` to `primitives/merkle-mountain-range` so `MmrApi` can be used by `client` code.

##

Fixes https://github.com/paritytech/grandpa-bridge-gadget/issues/42
Fixes https://github.com/paritytech/grandpa-bridge-gadget/issues/406

##

Polkadot companion: https://github.com/paritytech/polkadot/pull/5276
Cumulus companion: https://github.com/paritytech/cumulus/pull/1161

### CODE REVIEW

File: client/beefy/src/worker.rs [hunk 1]
```diff
@@ -550,37 +520,17 @@ where
...
-
-	/// For tests, have the option to modify mmr root.
-	#[cfg(test)]
-	fn extract_mmr_root_digest(&self, header: &B::Header) -> Option<MmrRootHash> {
-		let mut mmr_root = find_mmr_root_digest::<B>(header);
-		if self.test_res.corrupt_mmr_roots {
-			mmr_root.as_mut().map(|hash| *hash ^= MmrRootHash::random());
-		}
-		mmr_root
+	/// Simple wrapper over mmr root extraction. Tries from header digest, then from client state.
```
Comment (svyatonik) [inline client/beefy/src/worker.rs:523]: I think this comment is obsolete - the function now only uses runtime api :)

File: frame/merkle-mountain-range/primitives/src/lib.rs [hunk 2]
```diff
@@ -429,6 +429,9 @@ sp_api::decl_runtime_apis! {
 		/// The leaf data is expected to be encoded in it's compact form.
 		fn verify_proof_stateless(root: Hash, leaf: EncodableOpaqueLeaf, proof: Proof<Hash>)
 			-> Result<(), Error>;
+
+		/// Return the on-chain MMR root hash.
+		fn mmr_root() -> Result<Hash, Error>;
```
Comment (acatangiu) [inline frame/merkle-mountain-range/primitives/src/lib.rs:434]: Does adding this new fn to the runtime api require to also version the api declaration (`#[api_version(x)]`)?

**Review by svyatonik — APPROVED** [review]
Looks good!

**Comment by acatangiu** [comment]
Moved MMR primitives from `frame/merkle-mountain-range/primitives` to `primitives/merkle-mountain-range` so `MmrApi` can be used by client code.

File: frame/merkle-mountain-range/src/lib.rs [hunk 3]
```diff
@@ -70,7 +70,29 @@ mod mock;
...
+/// so that any point in time in the future we can receive a proof about some past
+/// blocks without using excessive on-chain storage.
+///
+/// Hence we implement the [LeafDataProvider] for [ParentNumberAndHash] which is a
+/// crate-local wrapper over [frame_system::Pallet]. Since the current block hash
+/// is not available (since the block is not finished yet),
+/// we use the `parent_hash` here along with parent block number.
+pub struct ParentNumberAndHash<T: frame_system::Config> {
+	_phanthom: sp_std::marker::PhantomData<T>,
+}
```
Comment (andresilva) [inline frame/merkle-mountain-range/src/lib.rs:85]:
```suggestion
}

```

**Review by andresilva — COMMENTED** [review]
IMO the worker should still rely on digests for fetching the MMR root. Doing it with a runtime call will make it harder for any future light client to verify BEEFY (since it requires a runtime call which a light client can't do), and this would also require changes to state pruning since validators will prune state on finality in which case the runtime call to fetch the MMR root will fail.

**Comment by andresilva** [comment]
To be clear, I think adding the runtime API and all the other changes in this PR are good, but due to the reasons above I think it would be safer to keep using the digest.

File: frame/merkle-mountain-range/primitives/src/lib.rs (hunk 2 above)
Comment (andresilva) [reply frame/merkle-mountain-range/primitives/src/lib.rs:419]: I think that shouldn't be necessary.

**Comment by andresilva** [comment]
If we really want to fetch the MMR root from storage then we need to do it ASAP as explained in https://github.com/paritytech/grandpa-bridge-gadget/issues/42, i.e. listen to block import notifications extract the MMR root and somehow persist them so that they can be used later the gadget when it's time to vote, otherwise the state won't be there at the time of voting.

**Comment by acatangiu** [comment]
> IMO the worker should still rely on digests for fetching the MMR root. Doing it with a runtime call will make it harder for any future light client to verify BEEFY (since it requires a runtime call which a light client can't do), and this would also require changes to state pruning since validators will prune state on finality in which case the runtime call to fetch the MMR root will fail.

It's my understanding that light clients wouldn't even need to sync Substrate headers, but could work exclusively using BEEFY payload, they have the MMR root hash in the BEEFY payload.

> If we really want to fetch the MMR root from storage then we need to do it ASAP as explained in https://github.com/paritytech/grandpa-bridge-gadget/issues/42, i.e. listen to block import notifications extract the MMR root and somehow persist them so that they can be used later by the gadget when it's time to vote, otherwise the state won't be there at the time of voting.

Hmm.. so you're saying (Tomek says the same in issue) that after block is GRANDPA finalized, the state for that block can be pruned, I see. I've not seen this in my tests, but there's multiple knobs at work here probably and in my simple local tests I didn't hit the fatal combination.

Ok, in this case, I'll change current PR to:
1. save MMR root to state
2. keep getting MMR roots from header digests for now

then we can incrementally make the switch for the worker/voter to read from state on block import (and maybe also remove `MmrRoot` header digest altoghether).

**Review by andresilva — APPROVED** [review]

**Comment by acatangiu** [comment]
bot merge

**Comment by acatangiu** [comment]
bot merge

**Comment by TriplEight** [comment]
bot merge

**Comment by TriplEight** [comment]
bot merge

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
bot merge force

**Comment by acatangiu** [comment]
bot merge

**Pull request merged**